src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs

```diff
@@ -158,7 +161,15 @@ public virtual void OnSocketClosed()
         // Called on Libuv thread
         public void Tick()
         {
-            _frame.Tick();
+            if (_timeoutEnabled)
+            {
+                if (_secondsSinceLastConnectionEvent > ServerOptions.Limits.ConnectionTimeout.TotalSeconds)
+                {
+                    StopAsync();
+                }
+
+                _secondsSinceLastConnectionEvent++;
```

Comment:
This relies on `Tick()` being called exactly once per second, which seems fragile.  Have you compared the perf of something like `System.Diagnostics.Stopwatch()`, which would basically track the precise time of the last connection event, rather than an integer number of seconds?